### DISCUSSION TITLE
Deallocation semantics and problems with `Vector`

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#8479] [General]
We have discussed the semantics of regions and have talked about deallocation semantics - once your region is out of scope, all memory in that region is deallocated.

This understanding of regions would break `Vector`, look at this function
```
pub def empty(): Vector[a] = region rc {
        let arr = Array#{} @ rc;
        toVector(arr)
    }

pub def toVector(a: Array[a, r]): Vector[a] \ r = region rc {
        let arr1 = copyOfRange(rc, 0, length(a), a);
        unchecked_cast(arr1 as Vector[a])
    }
```
Now your newly created vector is instantly a dangling pointer.

Is this a problem with Vector or a problem of the deallocation semantics? I also want to point out that the same problem occurs if you want to write a Lazy wrapper in Flix, or other structures that are properly encapsulated mutable memory.

**Comment by magnus-madsen** [comment]
I don't understand.

**Comment by JonathanStarup** [comment]
@magnus-madsen I've clarified my thoughts

**Comment by magnus-madsen** [comment]
The example just shows that copying from Array to Vector should be an intrinsic. There is no fundamental problem (?). I can write the same function for Array and List without issue.

**Comment by JonathanStarup** [comment]
i think it shouldnt be intrinsic. An implementation of Lazy would need the same things. and any other pure interface with encapsulated mutation. 

i think its something people will need in general, not just a one off we can ad hoc
